### PR TITLE
Improve BitVecOps<>::Iter::NextElem

### PR DESCRIPTION
**Pull request by BruceForstall**
Tweak the implementation, to reduce the number of instructions
executed in the hot path.

Also, don't pass "env" to NextElem; it can be stored by Init()
if required. For non-inlined calls, this saves setting up one
argument.

Overall, this reduces instruction count of superpmi over
a minopts test run of x86 release by 2.6% (NextElem is very hot).

### CODE REVIEW

**Comment by BruceForstall** [comment]
@CarolEidt @pgavlin PTAL
cc @dotnet/jit-contrib

File: src/jit/bitsetasshortlong.h [hunk 1]
```diff
@@ -467,24 +467,23 @@ class BitSetOps</*BitSetType*/ BitSetShortLongRep,
         unsigned m_bitNum; // The number of bits that have already been iterated over (set or clear).  If you
         // add this to the bit number of the next bit in "m_bits", you get the proper bit number of that
         // bit in "m_bs".
+        Env m_env;
```
Comment (briansull) [inline src/jit/bitsetasshortlong.h]: removing - That pesky undocumented, unused "env" parameter.

File: src/jit/bitset.h [hunk 2]
```diff
@@ -442,10 +442,10 @@ class BitSetOpsWithCounter
         {
         }
 
-        bool NextElem(Env env, unsigned* pElem)
+        bool NextElem(unsigned* pElem)
         {
             BitSetTraits::GetOpCounter(env)->RecordOp(BitSetSupport::BSOP_NextBit);
```
Comment (briansull) [inline src/jit/bitset.h]: This is the compilation error, a reference to "env"

File: src/jit/bitset.h [hunk 3]
```diff
@@ -271,7 +271,7 @@ class BitSetOps
     class Iter {
       public:
         Iter(Env env, BitSetValueArgType bs) {}
```
Comment (briansull) [inline src/jit/bitset.h]: Reference to "env"

File: src/jit/bitsetasshortlong.h (hunk 1 above)
Comment (briansull) [inline src/jit/bitsetasshortlong.h]: Can you add a comment here.

File: src/jit/bitsetasuint64.h [hunk 4]
```diff
@@ -213,8 +213,11 @@ class BitSetOps</*BitSetType*/ UINT64,
         {
         }
 
-        bool NextElem(Env env, unsigned* pElem)
+        bool NextElem(unsigned* pElem)
         {
+            // TODO-Throughtput: use BitScanForward64() intrinsic (see short/long implementation).
+            // REVIEW-Bug: This assumes *pElem comes in as the last bit numbered returned, or zero to start.
+            // This doesn't seem right; the current position should be carried with the iterator.
```
Comment (CarolEidt) [inline src/jit/bitsetasuint64.h]: Are you going to open an issue for this?

File: src/jit/bitsetasshortlong.h [hunk 5]
```diff
@@ -505,15 +504,12 @@ class BitSetOps</*BitSetType*/ BitSetShortLongRep,
...
-                    m_bitNum += nextBit + 1;
-                    m_bits >>= nextBit;
-                    m_bits >>= 1; // Have to do these separately -- if we have 0x80000000, nextBit == 31, and shifting
-                                  // by 32 bits does nothing.
+                    m_bits &= ~(((size_t)1) << nextBit); // clear bit we just found so we don't find it again
                     return true;
                 }
                 else
                 {
-                    unsigned len = BitSetTraits::GetArrSize(env, sizeof(size_t));
```
Comment (pgavlin) [inline src/jit/bitsetasshortlong.h]: For a further improvement you could hoist this call into the constructor and replace `m_env` with `m_len`.

File: src/jit/bitsetasuint64.h (hunk 4 above)
Comment (BruceForstall) [inline src/jit/bitsetasuint64.h]: I went ahead and just fixed it.

**Comment by BruceForstall** [comment]
@dotnet-bot test this please

**Pull request merged**